velox/dwio/common/BitUnpacking.h

```diff
@@ -0,0 +1,574 @@
+
+  switch (bitWidth) {
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+      unpack5to8(bitWidth, inputBuffer, numValues, outputBuffer);
+      break;
+    case 8:
+      unpack8_cast(inputBuffer, numValues, outputBuffer);
+      break;
+    case 9:
+    case 11:
+    case 13:
+    case 15:
+    case 10:
+    case 12:
+    case 14:
+      unpack9to15(bitWidth, inputBuffer, numValues, outputBuffer);
+      break;
+    case 16:
+      unpack16(inputBuffer, numValues, outputBuffer);
+      break;
+    default:
+      DWIO_RAISE("invalid bitWidth");
+  }
+}
+
+/// Unpack numValues number of uint8_t values from inputBuffer. The results will
+/// be written to outputBuffer. numValues must be a multiple of 8. The caller
+/// needs to make sure the inputBufferLen contains at least numValues number of
+/// packed values.
+inline void unpack(
+    uint8_t bitWidth,
+    const uint8_t*& inputBuffer,
+    uint64_t inputBufferLen,
+    uint64_t numValues,
+    uint32_t*& outputBuffer) {
+  DWIO_ENSURE(bitWidth >= 1 && bitWidth <= 32);
+  DWIO_ENSURE((numValues & 0x7) == 0);
+  DWIO_ENSURE(inputBufferLen * 8 >= bitWidth * numValues);
+
+  switch (bitWidth) {
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+      unpack1to7(bitWidth, inputBuffer, numValues, outputBuffer);
+      break;
+    case 8:
+      unpack8(inputBuffer, numValues, outputBuffer);
+      break;
+    case 9:
+    case 10:
+    case 11:
+    case 12:
+    case 13:
+    case 14:
+    case 15:
+      unpack9to15(bitWidth, inputBuffer, numValues, outputBuffer);
+      break;
+    case 16:
+      unpack16(inputBuffer, numValues, outputBuffer);
+      break;
+    case 17:
+    case 18:
+    case 19:
+    case 20:
+    case 21:
+    case 22:
+    case 23:
+    case 24:
+    case 25:
+    case 26:
+    case 27:
+    case 28:
+    case 29:
+    case 30:
+    case 31:
+    case 32:
+      VELOX_NYI(
+          "Unpacking 32 bit integers of bit width in [17, 32] not implemented");
+    default:
+      DWIO_RAISE("invalid bitWidth");
+  }
+}
+
+void generatePdepMasks(uint8_t outBitWidth) {
+  uint8_t numBatches = 64 / outBitWidth;
+  for (uint8_t i = 0; i <= outBitWidth; i++) {
+    uint64_t mask = (1l << i) - 1;
+    uint64_t val = mask;
+    for (uint8_t j = 1; j < numBatches; j++) {
+      val = (val << outBitWidth) | mask;
+    }
+    std::cout << "0x" << std::hex << std::setfill('0') << std::setw(16) << val
+              << ", ";
+  }
+  std::cout << std::endl;
+}
+} // namespace facebook::velox::dwio::common
+
+#else
+namespace facebook::velox::dwio::common {
+
+inline void unpack(
+    uint8_t bitWidth,
+    const uint8_t*& inputBuffer,
+    uint64_t inputBufferLen,
+    uint64_t numValues,
+    uint8_t*& outputBuffer) {
+  DWIO_ENSURE(bitWidth >= 1 && bitWidth <= 8);
+  DWIO_ENSURE((numValues & 0x7) == 0);
+  DWIO_ENSURE(inputBufferLen * 8 >= bitWidth * numValues);
+
+  VELOX_NYI();
```

Comment:
Maybe have some fallback implementation here (does not need to be optimized) so that we can start using them in the codebase

Reply:
Added the missing functions for bitWidth 17 to 32.